Don't show select cursor for any already selected actors and rework classic mouse style shift-selection

matjaeck opened the pull request:
Fixes #17313.
Fixes #17041.

This also includes a fix for classic mouse style when adding or removing actors from the current selection. Holding Shift+Ctrl while pointing at own actors will for classic mouse style always select unselected actors and deselect selected actors, using the select cursor. When not holding Shift, the default cursor is used when pointing at selected actors. The behavior for default mouse orders remains the same.

Added a fix for #17041 as it was needed to make shift-selecting actors work properly for spectators. I also used the opportunity to simplify some of the code and remove confusing usage of `LocalPlayer` to determine who is the viewer, since it can be done based on what we know about the `RenderPlayer`.

reaperrr commented:
We should consider picking at least the first commit to prep, assuming it's trivial enough to not require another playtest (though maybe a devtest).

matjaeck commented:
Removed the [changes](https://github.com/OpenRA/OpenRA/compare/bleed...matjaeck:fix-modify-selection) from the second commit for now as they break queuing resupply orders.

pchote commented:
My local experimentation (which I ran out of time to fully test / PR) showed that this fix needs to happen inside `InputOverridesSelection`, otherwise we change the cursor but still drop the selection when the player clicks.

matjaeck commented:
> My local experimentation (which I ran out of time to fully test / PR) suggested that this fix needs to happen inside `InputOverridesSelection`, otherwise we change the cursor but still drop the selection when the player clicks.

Sorry for asking, but do you talk about the changes that I have removed or about the remaining change?

matjaeck commented:
Updated description:
> Updated with a fix for adding or removing actors from the current selection. To avoid any conflicts with queuing orders, the "modify-selection-mode" is now activated by holding down Shift+Ctrl while pointing at an own selectable actor. The select cursor is used for both selecting and deselecting, like in the original RA. For RMB-setup, any orders are suppressed while selection-modifying is active, to avoid accidentally giving a force-attack order.

I took this approach because I came to the conclusion that we can no longer use Shift for the feature because of the conflict with queuing orders. Utilizing Shfit+Ctrl for it seemed to be least invasive (banning queued force-attack orders vs own actors should not be a problem) and the modifier-combination works well with assigning teams.

If the general approach here is acceptable, I'd welcome some help to fix the potential problems with the Ctrl-modifier on macOS.

tovl reviewed (changes requested):
It's a bad idea to gimp the default controls to fix a minor issue that only exists with the classic controls. shift-deselecting of units worked fine before with RMB orders and there is no added value in preventing orders being given in these circumstances&mdash;there is no reasonable risk of accidentally ordering when trying to select because with this set-up LMB is always selecting and RMB is always ordering.

I also think we should try to avoid needing multiple modifier keys for a single actual modifier unless there is absolutely no other alternative available.

In this case I think a better solution to the problem  would be to use shift + RMB for partial deselection when in classic mouse mode and to leave the default mouse mode alone. This doesn't conflict with anything else and is actually consistent with the general pattern in the classic mouse mode that RMB always deselects.

matjaeck commented:
> I also think we should try to avoid needing multiple modifier keys for a single actual modifier unless there is absolutely no other alternative available.

I can't see other options to support both queuing and selection-modifying.

> In this case I think a better solution to the problem would be to use shift + RMB for partial deselection when in classic mouse mode and to leave the default mouse mode alone. 

This does not solve the problem of adding an APC to a selection with infantry.

PS: I also use the default RMB orders and hope we can avoid any changes to hotkeys, but don't see a way to support LMB selection-modifying without.

matjaeck commented:
Updated. The force-selection-mode now only applies to classic mouse style while default mouse orders remain untouched. It's a bit disappointing that in the default mode shift-selection is not indicated by the proper cursor I think but I'm not sure if changing this to show the select cursor while allowing orders is any better.

tovl commented:
In the later c&c games hovering the mouse over a friendly unit would always show the selection cursor unless a force modifier is applied. right clicking would still issue a move order despite showing the selection cursor. This might not be a bad idea.

Commit:
Introduce force selection for classic mouse style.
Holding Shift+Ctrl while pointing at own actors will for classic mouse style always select unselected actors and deselect selected actors

Commit:
Overhaul cursors for shift-selecting actors.

Commit:
Partly refactor WICW and limit shift-selecting to viewer's actors.
Shift-selecting is generally allowed for spectators in everyone's view
or when shroud is disabled.

matjaeck commented:
Updated again. The select cursor is shown when shift is held while pointing at own actors. For classic mouse style, force-selection is required. Always showing the select cursor when pointing at own actors hides useful order targeters like enter or deploy. Denying shift-selecting actors that are not your own makes sense IMO and it made it easier to achieve the desired cursor behavior.

Commit:
Use RenderPlayer to determine current viewer.

matjaeck commented:
Updated. Added a fix for #17041 as it was needed to make shift-selecting actors work properly for spectators. I also used the opportunity to simplify some of the code and remove confusing usage of `LocalPlayer` to determine who is the viewer, since it can be done based on what we know about the `RenderPlayer`.

tovl commented:
> Always showing the select cursor when pointing at own actors hides useful order targeters like enter or deploy.

`TargetOverridesSelection` is always true for these order targeters. It would make sense to just use `InputOverridesSelection` to determine whether to show the selection cursor for either mouse style. This seems to be the C&C3 behaviour anyway. Not sure if it is better or worse than your current proposal.

matjaeck commented:
> It would make sense to just use InputOverridesSelection to determine whether to show the selection cursor for either mouse style. This seems to be the C&C3 behaviour anyway.

I'll try this out on a new branch. In any case, your changes in #17099 should be prioritised over this. Perhaps I should and I probably will drop the changes for shift-selecting entirely and only keep the fixes for the mentioned issues, so that comptetent people can fix this properly.

matjaeck commented:
Closing, comments here indicated that this is not the right way to fix the issue and other people will be able to implement a better solution much quicker. The PR still has its use as a reference for the issue I opened above, so closing it in its current state instead of dropping the shift-selection changes seems the best thing to do.
